File: android/app/src/main/java/com/francescooddo/buddygrammar/core/HandwritingTextFormatter.kt
package com.francescooddo.buddygrammar.core

/**
 * Applies English-only casing cleanup to recognized handwriting. Recognition
 * in every other language is preserved verbatim so noun and script-specific
 * casing rules are not overwritten by English heuristics.
 */
object HandwritingTextFormatter {
    fun textForInsertion(
        recognizedText: String,
        contextBeforeCursor: String?,
        languageTag: String = LanguageSupport.DEFAULT_LANGUAGE_TAG,
    ): String {
        // Unknown editor context cannot justify changing recognized casing.
        if (contextBeforeCursor == null) return recognizedText
        val usesEnglishRules = LanguageSupport.usesEnglishPriors(languageTag)
        if (!usesEnglishRules) return recognizedText
        var text = recognizedText

        val letters = text.filter { it.isLetter() }
        val isAllCaps = letters.length > 1 &&
            letters.any { it.isUpperCase() } &&
            letters.none { it.isLowerCase() }
        if (isAllCaps) text = text.lowercase()

        text = text.replace(Regex("\\bi\\b"), "I")

        val firstLetter = text.indexOfFirst { it.isLetter() }
        if (firstLetter == -1) return text

        if (SuggestionEngine.isSentenceStart(contextBeforeCursor)) {
            text = text.replaceRange(
                firstLetter,
                firstLetter + 1,
                text[firstLetter].uppercaseChar().toString(),
            )
        } else if (shouldLowercaseFirstWord(text, firstLetter)) {
            text = text.replaceRange(
                firstLetter,
                firstLetter + 1,
                text[firstLetter].lowercaseChar().toString(),
            )
        }
        return text
    }

    /**
     * Demotes a leading capital in the middle of a sentence, but only for
     * plain Title-case words. "I", acronyms ("NASA"), and mixed-case words
     * ("BuddyGrammar", "iPhone") keep their casing.
     */
    private fun shouldLowercaseFirstWord(text: String, firstLetter: Int): Boolean {
        if (!text[firstLetter].isUpperCase()) return false
        val word = text.substring(firstLetter).takeWhile { it.isLetter() || it == '\'' }
        if (word == "I" || word.startsWith("I'")) return false
        return word.drop(1).none { it.isUpperCase() }
    }
}

File: BuddyGrammarKit/Sources/BuddyGrammarKit/HandwritingTextFormatter.swift
import Foundation

public enum HandwritingTextFormatter {
    public static func textForInsertion(
        _ recognizedText: String,
        contextBeforeInput: String?,
        languageCode: String? = nil
    ) -> String {
        guard LanguageSupport.usesEnglishPriors(languageCode: languageCode) else {
            return recognizedText
        }
        var text = recognizedText

        let letters = text.filter(\.isLetter)
        let isAllCaps = letters.count > 1
            && letters.contains(where: \.isUppercase)
            && !letters.contains(where: \.isLowercase)
        if isAllCaps {
            text = text.lowercased()
        }

        text = text.replacingOccurrences(
            of: #"\bi\b"#,
            with: "I",
            options: .regularExpression
        )

        guard let firstLetter = text.firstIndex(where: \.isLetter) else {
            return text
        }

        if TypingContextAnalyzer.analyze(contextBeforeInput).isAtSentenceStart {
            text.replaceSubrange(
                firstLetter...firstLetter,
                with: String(text[firstLetter]).uppercased()
            )
        } else if shouldLowercaseFirstWord(of: text, startingAt: firstLetter) {
            text.replaceSubrange(
                firstLetter...firstLetter,
                with: String(text[firstLetter]).lowercased()
            )
        }
        return text
    }

    /// Demotes a leading capital in the middle of a sentence, but only for
    /// plain Title-case words. "I", acronyms ("NASA"), and mixed-case words
    /// ("BuddyGrammar", "iPhone") keep their casing.
    private static func shouldLowercaseFirstWord(
        of text: String,
        startingAt firstLetter: String.Index
    ) -> Bool {
        guard text[firstLetter].isUppercase else { return false }
        let word = String(text[firstLetter...].prefix { $0.isLetter || $0 == "'" })
        guard word != "I", !word.hasPrefix("I'") else { return false }
        return word.dropFirst().allSatisfy { !$0.isUppercase }
    }
}
